TYP: Add several missing return type annotations

QuLogic opened the pull request:
## PR summary

I'm probably least confident in the `backend_bases.pyi` types as it's a bit of a complex file.

One that is missing is `Affine2DBase.is_separable`, because adding the return annotation makes it typed, and then `mypy` complains that:
```
lib/matplotlib/transforms.pyi:229: error: Cannot override writeable attribute with read-only property  [override]
```

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

QuLogic commented:
Depends on #26450 to fix the subclass method inconsistency.

ksunden commented:
Cycling to get the mypy runs rebased post merge of #26450

ksunden commented:
`is_separable` and a fair number of other things like it fall into the category of things that are really intended to be read-only, but are not wholly implemented as such (but sometimes are?)...


These were problematic for stubtest, in ways that were more trouble than initially worth it, so I ignored the noisy ones:

See https://github.com/matplotlib/matplotlib/blob/7484a86cddbbb881576377328861389a36d52361/ci/mypy-stubtest-allowlist.txt#L35-L45

All in all, not a _very_ user-facing (if technically public, and useful in some contexts) set of variables, and setting them would wreak havoc.

ksunden commented on lib/matplotlib/tri/_tripcolor.pyi:
```diff
@@ -17,10 +18,10 @@ def tripcolor(
     vmin: float | None = ...,
     vmax: float | None = ...,
-    shading: Literal["flat", "gouraud"] = ...,
+    shading: Literal["flat"],
```
```suggestion
    shading: Literal["flat"] = ...,
```

ksunden replied:
This seems to be what stubtest is complaining about

ksunden commented on lib/matplotlib/tri/_tripcolor.pyi:
```diff
@@ -33,7 +34,38 @@ def tripcolor(
     vmin: float | None = ...,
     vmax: float | None = ...,
-    shading: Literal["flat", "gouraud"] = ...,
+    shading: Literal["flat"],
```
```suggestion
    shading: Literal["flat"] = ...,
```

ksunden commented on lib/matplotlib/text.pyi:
```diff
@@ -103,9 +103,9 @@ class Text(Artist):
     def set_parse_math(self, parse_math: bool) -> None: ...
     def get_parse_math(self) -> bool: ...
-    def set_fontname(self, fontname: str | Iterable[str]): ...
+    def set_fontname(self, fontname: str | Iterable[str]) -> None: ...
```
Maybe should remove the `return ` from the implementation, and maybe call the canonically defined version of `set_fontfamily` as `set_family` is not type checkable as an alias.

ksunden commented on lib/matplotlib/spines.pyi:
```diff
@@ -63,7 +63,7 @@ class Spine(mpatches.Patch):
 class SpinesProxy:
     def __init__(self, spine_dict: dict[str, Spine]) -> None: ...
-    def __getattr__(self, name: str): ...
+    def __getattr__(self, name: str) -> Callable[[Any], None]: ...
```
```suggestion
    def __getattr__(self, name: str) -> Callable[[...], None]: ...
```

May work with `Any`, but typing docs indicate `...` as the "any input signature" syntax

QuLogic replied:
I originally thought it was only _one_ parameter with any type, as passed to any `.set_*` method, but now I realize the proxy can be for `.set`, which allows multiple parameters, so `...` is better.

ksunden commented on lib/matplotlib/mathtext.pyi:
```diff
@@ -14,13 +14,8 @@ from matplotlib.typing import ColorType
-        antialiased: bool | None = ...
-    ): ...
+        self, s: str, dpi: float = ..., prop: FontProperties | None = ..., *, antialiased: bool | None = ...
+    ) -> RasterParse | VectorParse: ...
```
Wish there was a better way here... a _given_ MathTextParser knows which it will return, but that is a state of the mathtext parser, not of this method...

Could maybe do some funky generic things with `__new__`? (I'd say `__init__`, but don't have access to the type of the constructed object there) but that gets weird quickly, so probably not worth it.

QuLogic replied:
Something like this seems to work if we want to apply it:
```patch
diff --git a/lib/matplotlib/mathtext.pyi b/lib/matplotlib/mathtext.pyi
index 706f83a6b1..1b0468d59a 100644
--- a/lib/matplotlib/mathtext.pyi
+++ b/lib/matplotlib/mathtext.pyi
@@ -1,5 +1,8 @@
 import os
+from typing import Generic, IO, Literal, TypeVar, overload
+
 from matplotlib.font_manager import FontProperties
+from matplotlib.typing import ColorType

 # Re-exported API from _mathtext.
 from ._mathtext import (
@@ -8,14 +11,17 @@ from ._mathtext import (
     get_unicode_index as get_unicode_index,
 )

-from typing import IO, Literal
-from matplotlib.typing import ColorType
+_ParseType = TypeVar("_ParseType", RasterParse, VectorParse)

-class MathTextParser:
-    def __init__(self, output: Literal["path", "agg", "raster", "macosx"]) -> None: ...
+class MathTextParser(Generic[_ParseType]):
+    @overload
+    def __init__(self: MathTextParser[VectorParse], output: Literal["path"]) -> None: ...
+    @overload
+    def __init__(self: MathTextParser[RasterParse], output: Literal["agg", "raster", "macosx"]) -> None: ...
+    def __init__(self: MathTextParser[_ParseType], output: Literal["path", "agg", "raster", "macosx"]) -> None: ...
     def parse(
         self, s: str, dpi: float = ..., prop: FontProperties | None = ..., *, antialiased: bool | None = ...
-    ) -> RasterParse | VectorParse: ...
+    ) -> _ParseType: ...

 def math_to_image(
     s: str,
```

QuLogic replied:
Though it seems like we'd need to add extra casts on usage to specify the generic, so it's maybe not so great?

ksunden commented on lib/matplotlib/collections.pyi:
```diff
@@ -84,8 +84,8 @@ class PathCollection(_CollectionWithSizes):
-        **kwargs
-    ): ...
+        **kwargs,
+    ) -> tuple[list[Artist], list[str]]: ...
```
```suggestion
    ) -> tuple[list[Line2D], list[str]]: ...
```

According to the docstring this _can_ be more specific.

Would require an import as well.

ksunden commented on lib/matplotlib/backend_bases.pyi:
```diff
@@ -76,7 +76,7 @@ class RendererBase:
         antialiased: bool,
         edgecolors: Sequence[ColorType] | ColorType | None,
-    ): ...
+    ) -> None: ...
```
Slightly odd that the implementation has a `return ` statement, but I do agree that `draw_path_collection` also returns `None`, so I guess we _should_ propegate that. Also just consistent with other `draw_` methods

ksunden commented on lib/matplotlib/__init__.pyi:
```diff
@@ -96,7 +96,7 @@ def rc_file(
 @contextlib.contextmanager
 def rc_context(
     rc: dict[str, Any] | None = ..., fname: str | Path | os.PathLike | None = ...
-): ...
+) -> Generator[None, None, None]: ...
```
Should we be doing `AbstractContextManager` here instead of the decorator?

QuLogic replied:
If I do that, then mypy complains that `AbstractContextManager[None]` is not callable when used as a decorator. I wonder if we should revert the others as well.

QuLogic commented:
Fixed all those items, and also added return type to `_LazyTickList`, which seems to have caused some new errors in `test_axes.py`, but those are just accessing private attributes.
